crates/ra_syntax/src/ast/tokens.rs

```diff
@@ -172,3 +172,132 @@ impl RawString {
         Some(range + contents_range.start())
     }
 }
+
+#[derive(Debug)]
+pub enum StringPiece {
+    Quote,
+    LiteralSequence,
+    EscapeSequence,
+    FormatSpecifier,
+}
+
+impl String {
+    pub fn lex<F>(&self, is_fmt_str: bool, callback: &mut F)
+    where
+        F: FnMut(TextRange, StringPiece),
+    {
+        let src = self.text().as_str();
+        let initial_len = src.len();
+        let mut chars = src.chars();
+
+        while let Some(first_char) = chars.next() {
+            let start = initial_len - chars.as_str().len() - first_char.len_utf8();
+            let piece = match first_char {
+                '\\' => {
+                    // escape secquence, see https://doc.rust-lang.org/reference/tokens.html
+                    match chars.clone().next() {
+                        Some('\\') | Some('\n') | Some('n') | Some('r') | Some('t') | Some('0')
+                        | Some('\'') | Some('\"') => {
+                            chars.next();
+                            StringPiece::EscapeSequence
+                        }
+                        Some('x') => {
+                            // ascii escape
+                            chars.next();
+                            let mut cloned = chars.clone().take(2);
+                            let mut piece = None;
+                            while let Some(next_char) = cloned.next() {
+                                match next_char {
+                                    '0'..='F' => {
+                                        chars.next();
+                                    }
+                                    _ => {
+                                        piece = Some(StringPiece::LiteralSequence);
+                                        break;
+                                    }
+                                }
+                            }
+                            piece.unwrap_or(StringPiece::EscapeSequence)
+                        }
+                        Some('u') => {
+                            // unicode escape
+                            chars.next();
+                            let mut cloned = chars.clone().take(8); // up to 6 digits + opening `{` and closing `}`
+                            if let Some(next_char) = cloned.next() {
```

Comment:
Sorry, I wasn't clear about this, we already have logic for this implemented in rustc_lexer:

https://github.com/rust-lang/rust/blob/52fa23add6fb0776b32cc591ac928618391bdf41/src/librustc_lexer/src/unescape.rs#L207-L260

So, rather than re-doing this, we should use the `unscape_xxx` family of funcitions.

Reply:
The unescape function cannot be used in a compositional way together with the format specifier so I decided to focus in this PR on the format specifier. In a follow-up it should be easy to include escape sequences by using the unescape functions and a second iteration over the string as escape sequences and format specifier should have mutual exclusive ranges (and the sorting of syntax ranges was implemented in #4022).